### PR TITLE
FEValues: Allow checking for cell similarities in multi-threaded mode

### PR DESCRIPTION
**Pull request by masterleinad**
Looking at the `Kokkos` `MatricxFree` benchmark, I noticed that the setup takes a long time without setting `DEAL_II_NUM_THREADS=1`. It seems that not checking for cell similarity is responsible. At the moment, we only allow doing that check when we set the number of threads to 1. There might be contexts where we know that we would not run into the danger of initializing `FEValues` objects by multiple threads (and getting possibly slightly different results depending on which thread looked at which cell).
This pull request introduces an option to explicitly allow this check and enables it in the `Kokkos` `MatrixFree` implementation.

### CODE REVIEW

**Review by Rombur — APPROVED** [review]

**Review by kronbichler — APPROVED** [review]
This looks good to me. Maybe you want to copy some of the comment in `fe_values_base.cc` to the new flag, such that the user can decide what enabling the flags means and why it is off by default.

File: include/deal.II/fe/fe_values_base.h [hunk 1]
```diff
@@ -222,6 +222,18 @@ class FEValuesBase : public Subscriptor
    */
   virtual ~FEValuesBase() override;
 
+  /**
+   * Explicitly allow to check for cell similarity.
+   * The detection of simple geometries with CellSimilarity is sensitive to the
+   * first cell detected. When using multiple threads, each thread might get a
+   * thread local copy of the FEValues object that is initialized to the first
+   * cell the thread sees. As this cell might be differ between runs and number
```
Comment (bangerth) [inline include/deal.II/fe/fe_values_base.h:230]:
```suggestion
   * cell the thread sees. As this cell might be different between runs and number
```

**Review by bangerth — COMMENTED** [review]
Perhaps call this function `always_allow_check_for_cell_similarity()` to make clear that it is already allowed under certain circumstances?

**Comment by masterleinad** [comment]
@bangerth I addressed your comments.

**Review by kronbichler — APPROVED** [review]

**Comment by masterleinad** [comment]
@bangerth ping

**Review by kronbichler — APPROVED** [review]

**Comment by kronbichler** [comment]
Let us merge this now, we can always further improve documentation later.

**Comment by bangerth** [comment]
Thanks. Apologies for not getting to this quicker!

**Pull request merged**